paimon-core/src/main/java/org/apache/paimon/table/ExpireSnapshotsImpl.java

```diff
@@ -215,10 +204,15 @@ public int expireUntil(long earliestId, long endExclusiveId) {
             }
 
             Snapshot snapshot = snapshotManager.snapshot(id);
-            snapshotDeletion.cleanUnusedManifests(snapshot, skippingSet);
+            snapshotDeletion.cleanUnusedManifests(snapshot, skippingSet, false);
 
-            // delete snapshot last
-            snapshotManager.fileIO().deleteQuietly(snapshotManager.snapshotPath(id));
+            // move snapshot metadata to changelog
+            try {
+                snapshotManager.moveToChangelog(id);
+                snapshotManager.commitLongLivedChangelogLatestHint(id);
```

Comment:
Maybe we should not just move snapshot file.

We could introduce a new JSON file for changelog.

Reply:
Currently, I've not seen the requirement to introduce a new metadata structure for changelog. We can rewrite the content later when we have to later.

Reply:
If we have to do it later, it is hard to keep forward compatibility.

It is better to provide a simple version now.